crates/matrix-sdk/src/widget/machine/driver_req.rs

```diff
@@ -253,6 +271,44 @@ impl FromMatrixDriverResponse for SendEventResponse {
     }
 }
 
+/// Ask the client to send a to-device message that corresponds to the given
+/// description.
+#[derive(Clone, Debug, Deserialize)]
+pub(crate) struct SendToDeviceRequest {
+    /// The type of the to-device message.
+    #[serde(rename = "type")]
+    pub(crate) event_type: String,
+    /// If the to-device message should be encrypted or not.
+    /// TODO: As per MSC 3819 should default to true
+    pub(crate) encrypted: bool,
+    /// The messages to be sent.
+    /// They are organized in a map of user ID -> device ID -> content like the
+    /// cs api request.
+    pub(crate) messages:
+        BTreeMap<OwnedUserId, BTreeMap<DeviceIdOrAllDevices, Raw<AnyToDeviceEventContent>>>,
+}
+
+impl From<SendToDeviceRequest> for MatrixDriverRequestData {
+    fn from(value: SendToDeviceRequest) -> Self {
+        MatrixDriverRequestData::SendToDeviceEvent(value)
+    }
+}
+
+impl MatrixDriverRequest for SendToDeviceRequest {
+    type Response = send_event_to_device::v3::Response;
+}
+
+impl TryInto<send_event_to_device::v3::Response> for MatrixDriverResponse {
+    type Error = anyhow::Error;
+
+    fn try_into(self) -> Result<send_event_to_device::v3::Response, Self::Error> {
+        match self {
+            MatrixDriverResponse::MatrixToDeviceSent(response) => Ok(response),
+            _ => Err(anyhow::Error::msg("bug in MatrixDriver, received wrong event response")),
```

Comment:
Let's use `serde::de::Error::custom()` instead. `anyhow` should generally not be used in libraries, it's meant to be used in applications.

A serde error makes sense here I would say.

Reply:
https://github.com/matrix-org/matrix-rust-sdk/pull/4963/commits/cca45ca63d98d909c9b701b450807dbd2c55cb0c